lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -9639,6 +9639,86 @@ namespace Js
+        {
+            FunctionBody* parentFuncBody = pfuncCaller->GetFunctionBody();
+            ModuleID id = parentFuncBody->GetModuleID();
+
+            if (!JavascriptString::Is(specifier)) {
+                JavascriptError::ThrowTypeError(scriptContext, JSERR_NeedString);
+            }
+
+            LPCOLESTR moduleName = Js::JavascriptString::FromVar(specifier)->GetSz();
+
+            HRESULT hr = 0;
+            if (id != kmodGlobal)
+            {
+                SourceTextModuleRecord *referenceModuleRecord = parentFuncBody->GetScriptContext()->GetLibrary()->GetModuleRecord(id);
+
+                ASYNC_HOST_OPERATION_START(scriptContext->GetThreadContext());
+                hr = scriptContext->GetHostScriptContext()->FetchImportedModule(referenceModuleRecord, moduleName, &moduleRecordBase);
+                ASYNC_HOST_OPERATION_END(scriptContext->GetThreadContext());
+            }
+            else
+            {
+                DWORD_PTR dwReferencingSourceContext = parentFuncBody->GetFunctionInfo()->GetFunctionProxy()->GetSourceContextInfo()->dwHostSourceContext;
+
+                ASYNC_HOST_OPERATION_START(scriptContext->GetThreadContext());
+                hr = scriptContext->GetHostScriptContext()->FetchImportedModuleFromScript(dwReferencingSourceContext, moduleName, &moduleRecordBase);
+                ASYNC_HOST_OPERATION_END(scriptContext->GetThreadContext());
+            }
+
+            if (FAILED(hr))
+            {
+                Js::JavascriptError * error = scriptContext->GetLibrary()->CreateURIError();
+                Var value = JavascriptString::NewCopySz(moduleName, scriptContext);
+                JavascriptOperators::OP_SetProperty(error, PropertyIds::message, value, scriptContext);
+                return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, error, scriptContext);
+            }
+
+            moduleRecord = SourceTextModuleRecord::FromHost(moduleRecordBase);
+
+            if (moduleRecord->WasEvaluated())
+            {
+                return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(true, moduleRecord->GetNamespace(), scriptContext, moduleRecord);
+            }
+            else if (moduleRecord->GetErrorObject() != nullptr)
+            {
+                return SourceTextModuleRecord::ResolveOrRejectDynamicImportPromise(false, moduleRecord->GetErrorObject(), scriptContext, moduleRecord);
+            }
+            else
+            {
+                JavascriptPromiseCapability *promiseCapability = moduleRecord->GetPromiseCapability();
+                if (promiseCapability == nullptr)
+                {
+                    BEGIN_TRANSLATE_TO_HRESULT(static_cast<ExceptionType>(ExceptionType_OutOfMemory | ExceptionType_StackOverflow));
+                    promiseCapability = JavascriptPromise::NewPromiseCapability(scriptContext->GetLibrary()->GetPromiseConstructor(), scriptContext);
```

Comment:
Does the spec call-out that we should be storing the `PromiseCapabilityRecord` here or is this more of an implementation detail? I would suspect it should be simpler to store a `JavascriptPromise*` here and call `JavascriptPromise::Resolve()` or `JavascriptPromise::Reject()` on it when it's time to resolve/reject it.

Looks like we're storing `promiseCapability` here to avoid casting the promise into a `JavascriptPromise*` because `JavascriptPromiseCapability::promise` is a `Var`? I think that cast should be safe to do since we pass `%Promise%` to `JavascriptPromise::NewPromiseCapability` which is guaranteed to return something that's a `JavascriptPromise*` because the return object is constructed via `JavascriptLibrary::CreatePromise` and we aren't doing a super call. Actually might be even simpler than that since we don't even need to call an executor function for this promise, right? So we could just call `library->CreatePromise()` and pass the result to `JavascriptPromise::InitializePromise`.

Might make sense to add a helper to `JavascriptPromise` itself

```c++
    JavascriptPromise* JavascriptPromise::CreateEnginePromise(ScriptContext* scriptContext)
    {
        // Unused
        JavascriptPromiseResolveOrRejectFunction* resolve;
        JavascriptPromiseResolveOrRejectFunction* reject;

        JavascriptPromise* promise = library->CreatePromise();
        InitializePromise(promise, &resolve, &reject, scriptContext);

        return promise;
    }
```

I think it would be simpler to resolve the promise at least if we stored the promise itself instead of the capability record. We would also avoid constructing a bunch of objects and making a bunch of calls by constructing the promise as the above helper.